README.md
# terraform-ssm-runcommand

This terraform module can be used to execute commands on remote EC2 instances via AWS SSM.

## Description

The module uses a `null_resource` resource which executes either a bash script or a powershell script depending on the detected OS.
The scripts are wrappers around [AWS CLI send-command](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/ssm/send-command.html), with some extra functionality.

## Usage

```terraform
module "ssm_runcommand_windows" {
  source                      = "github.com/paololazzari/terraform-ssm-runcommand"
  instance_id                 = "i-..."
  target_os                   = "windows"
  command                     = "Get-Process -name 'amazon*'"
}
```

```terraform
module "ssm_runcommand_unix" {
  source                      = "github.com/paololazzari/terraform-ssm-runcommand"
  instance_id                 = "i-..."
  target_os                   = "unix"
  command                     = "ps -ax | grep 'amazon*'"
}
```

For other examples, check the [examples](examples).

## Requirements

| Name | Version |
|------|---------|
| <a name="requirement_terraform"></a> [terraform](#requirement\_terraform) | >= 0.13.1 |
| <a name="requirement_aws"></a> [aws](#requirement\_aws) | >= 4.20.0 |
| <a name="requirement_null"></a> [null](#requirement\_null) | >= 3.2.1 |

## Providers

| Name | Version |
|------|---------|
| <a name="provider_aws"></a> [aws](#provider\_aws) | >= 4.20.0 |
| <a name="provider_null"></a> [null](#provider\_null) | >= 3.2.1 |

## Resources

| Name | Type |
|------|------|
| [null.ssm_runcommand_provisioner](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource) | resource |

## Modules

No modules.

## Inputs

The following inputs are required:

| Parameter Name              | Description                                                                            | Type     | Default | Required |
| ------------------          | -----------                                                                            | -------- | ------- | -------  |
| instance_id                 | The id of the EC2 instance on which to run the SSM command                             | string   |         | Yes      |
| target_os                   | The operating system of the EC2 instance on which to run the SSM command               | string   |         | Yes      |
| command                     | The command to execute on the EC2 instance                                             | string   |         | Yes      |

The following inputs are optional and can be used to control the behavior of the module:

| Parameter Name              | Description                                                                            | Type     | Default | Required |
| ------------------          | -----------                                                                            | -------- | ------- | -------- |    
| wait_for_command_completion | Whether or not the terraform execution should wait for the SSM command to be completed | bool     | false   | No       |
| ssm_timeout_seconds         | How many seconds the SSM agent has to start before the instance is deemed unhealthy    | int      | 60      | No       |
| show_command_output         | Whether or not the output of the SSM command should be printed                         | bool     | false   | No       | 
| continue_on_error           | Whether or not the terraform execution should continue if the SSM command failed       | bool     | false   | No       |
| log_file                    | If specified, any outputs will be redirected to it                                     | string   | ""      | No       |

The following inputs are optional and can be used to specify arguments for the send-command cli request:

| Parameter Name              | Description                                                                                             | Type     | Default  | Required |
| ------------------          | -----------                                                                                             | -------- | -------- | -------- | 
| timeout_seconds             | If this time is reached and the command hasn't already started running, it won't run                    | string   | ""       | No       |
| comment                     | User-specified information about the command, such as a brief description of what the command should do | string   | ""       | No       |
| output_s3_bucket_name       | The name of the S3 bucket where command execution responses should be stored                            | string   | ""       | No       |
| output_s3_key_prefix        | The directory structure within the S3 bucket where the responses should be stored                       | string   | ""       | No       |
| service_role_arn            | The ARN of the Identity and Access Management (IAM) service role to use to publish Amazon Simple Notification Service (Amazon SNS) notifications for Run Command commands          | string | "" | No       |
| notification_config         | Configurations for sending notifications                                                                | string   | ""       | No       |
| cloud_watch_output_config   | Enables Amazon Web Services Systems Manager to send Run Command output to Amazon CloudWatch Logs        | string   | ""       | No       |
| alarm_configuration         | The CloudWatch alarm you want to apply to your command                                                  | string   | ""       | No       |
| endpoint_url                | Override command's default URL with the given URL                                                       | string   | ""       | No       |
| region                      | The region to use. Overrides config/env settings                                                        | string   | ""       | No       |

For more information, see the [official documentation page](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/ssm/send-command.html).

## Outputs

No outputs.

CHANGELOG.md
# Changelog

<a name="v0.1.0"></a>
## [v0.1.0] - 2022-11-28

- Initial commit

util/ssm_run_command.sh
#!/bin/bash
instance_id=$1
target_os=$2
command=$3
wait_for_command_completion=$4
ssm_timeout_seconds=$5
show_command_output=$6
continue_on_error=$7
timeout_seconds=$8
comment=$9
output_s3_bucket_name=${10}
output_s3_key_prefix=${11}
service_role_arn=${12}
notification_config=${13}
cloud_watch_output_config=${14}
alarm_configuration=${15}
endpoint_url=${16}
region=${17}

[ -z "$1" ] && echo "instance_id is missing" && exit 1
[ -z "$2" ] && echo "target_os is missing" && exit 1
[ -z "$3" ] && echo "command is missing" && exit 1
[ -z "$4" ] && echo "wait_for_command_completion is missing" && exit 1
[ -z "$5" ] && echo "ssm_timeout_seconds is missing" && exit 1

# Validate configuration
validate () {
    # Validate AWS CLI installation
    aws --version >/dev/null 2>&1
    if [[ "$?" -ne 0 ]]; then
        echo "AWS CLI was not found"
        exit 1
    fi
}

# Build a string for extra arguments to pass to send-command api
build_extra_args() {
    extra_args=""

    declare -A extra_args_map=( 
        ["timeout-seconds"]="$timeout_seconds"
        ["comment"]="$comment"
        ["output-s3-bucket-name"]="$output_s3_bucket_name"
        ["output-s3-key-prefix"]="$output_s3_key_prefix"
        ["service-role-arn"]="$service_role_arn"
        ["notification-config"]="$notification_config"
        ["cloud-watch-output-config"]="$cloud_watch_output_config"
        ["alarm-configuration"]="$alarm_configuration"
        ["endpoint-url"]="$endpoint_url"
        ["region"]="$region"
    )

    for arg in "${!extra_args_map[@]}"; do
        if [ -n "${extra_args_map[$arg]}" ]; then
            extra_args="${extra_args} --$arg ${extra_args_map[$arg]}"
        fi
    done

    printf "%s" "$extra_args"
}

# Helper function for send-command requests
send_command () {
    # Choose the SSM document to run based on the OS of the target instance
    if [ "$target_os" == "windows" ]; then
        document_name="AWS-RunPowerShellScript"
    else
        document_name="AWS-RunShellScript"
    fi

    # Sanitize command
    sanitized_command=$(printf "%s" "$1" | sed 's/\$/\\$/g' | sed s/\'/"\\\'"/g)

    # Execute SSM command on target instance
    ssm_command="aws ssm send-command --instance-id $instance_id --document-name $document_name --parameters commands=\"'$sanitized_command'\" --query 'Command.CommandId' --output text $2"
    ssm_command_id=$(eval "$ssm_command" 2>&1)
    ssm_command_exit_code="$?"
    echo "$ssm_command_id"
    return "$ssm_command_exit_code"
}

# Wait for SSM to be up and running on the target EC2 instance
wait_ssm_bootstrap () {
    INSTANCE_STATE_ERR=$(cat <<-END
The instance is in an invalid state.
This can occur if:

- The instance ID is not valid
- The instance is not in the selected region
- The instance is not running
- The instance does not have the SSM agent running
- The instance does not have an instance profile with sufficient SSM permissions
END
)

  # Wait for SSM to be up and running for a duration defined by the user
  ssm_healthcheck_attempts=0
  ssm_healthcheck_max_attempts=$(expr "$ssm_timeout_seconds" / 10)

  while [[ "$ssm_healthcheck_attempts" -ne "$ssm_healthcheck_max_attempts" ]]
  do
    echo "Waiting for SSM to come online on instance $instance_id"
    ssm_command_output=$(send_command "ls" 2>&1)
    ssm_command_exit_code=$?
    if [[ "$ssm_command_exit_code" -eq 0 ]]; then
      echo "SSM is online on instance $instance_id"
      return
    else
      sleep 10
      ssm_healthcheck_attempts=$[$ssm_healthcheck_attempts +1]
    fi
  done

  echo "SSM did not come online in the specified time"
  
  # If the instance is not in a valid state, exit with a helpful message
  grep -qE "Instances .* not in a valid state" <<< "$ssm_command_output"
  if [[ "$?" -eq 0 ]]; then
    echo "$INSTANCE_STATE_ERR"
    exit 1
  else
    echo "Something went wrong."
    echo "$ssm_command_output"
    exit 1
  fi
}

# Run SSM command on the target EC2 instance
run_command () {
    # Get extra args
    extra_args=$(build_extra_args)

    # Execute command and make sure a valid command id is returned
    ssm_command_command_id=$(send_command "$command" "$extra_args")
    grep -qE "^[a-z0-9-]+$" <<< "$ssm_command_command_id"
    
    if [[ "$?" -ne 0 ]]; then
      echo "Something went wrong."
      echo "Output: $ssm_command_command_id"
      exit 1
    fi

    # If wait_for_command_completion is false, exit immediately
    if [ "$wait_for_command_completion" == "false" ]; then
      echo "SSM command has been executed"
      exit 0
    fi

    # If a different region was specified, include it when retrieving ssm command details
    if [ -n "${extra_args_map[region]}" ]; then
      region_arg="--region ${extra_args_map[region]}"
    fi

    # After the command has been executed, wait for its completion
    while true
    do
        ssm_get_command_invocation_stdout_command="aws ssm get-command-invocation --instance-id $instance_id --command-id $ssm_command_command_id --query 'StandardOutputContent' --output text $region_arg"
        ssm_get_command_invocation_stderr_command="aws ssm get-command-invocation --instance-id $instance_id --command-id $ssm_command_command_id --query 'StandardErrorContent' --output text $region_arg"
        ssm_command_invocation_stdout=$(eval $ssm_get_command_invocation_stdout_command)
        ssm_command_invocation_stderr=$(eval $ssm_get_command_invocation_stderr_command)

        if [[ "$ssm_command_invocation_stdout" == "" && "$ssm_command_invocation_stderr" == "" ]]; then
            : # If both STDOUT and STDERR are empty, then the command has not completed yet
        else
            if [[ "$ssm_command_invocation_stdout" != "" && "$show_command_output" == "true" ]]; then
                echo "SSM command standard output: $ssm_command_invocation_stdout"
            fi
            if [ -n "$ssm_command_invocation_stderr" ]; then
                if [ "$show_command_output" == "true" ]; then
                    echo "$ssm_command_invocation_stderr"
                fi

                if [ "$continue_on_error" == "false" ]; then
                    echo "SSM command execution had errors"
                    exit 1
                fi
            fi
            exit 0
        fi
    done
}

# Validate configuration
validate

# Wait for SSM to be up and running on target instance
wait_ssm_bootstrap

# Run SSM command on target instance
run_command

util/ssm_run_command.ps1
param(
    [parameter(Position=0,Mandatory=$true)][string] $instance_id,
    [parameter(Position=1,Mandatory=$true)][string] $target_os,
    [parameter(Position=2,Mandatory=$true)][string] $command,
    [parameter(Position=3,Mandatory=$true)][string] $wait_for_command_completion="false",
    [parameter(Position=4,Mandatory=$true)][string] $ssm_timeout_seconds="",
    [parameter(Position=5,Mandatory=$false)][string] $show_command_output="false",
    [parameter(Position=6,Mandatory=$false)][string] $continue_on_error="false",
    [parameter(Position=7,Mandatory=$false)][string] $timeout_seconds="",
    [parameter(Position=8,Mandatory=$false)][string] $comment="",
    [parameter(Position=9,Mandatory=$false)][string] $output_s3_bucket_name="",
    [parameter(Position=10,Mandatory=$false)][string] $output_s3_key_prefix="",
    [parameter(Position=11,Mandatory=$false)][string] $service_role_arn="",
    [parameter(Position=12,Mandatory=$false)][string] $notification_config="",
    [parameter(Position=13,Mandatory=$false)][string] $cloud_watch_output_config="",
    [parameter(Position=14,Mandatory=$false)][string] $alarm_configuration="",
    [parameter(Position=15,Mandatory=$false)][string] $endpoint_url="",
    [parameter(Position=16,Mandatory=$false)][string] $region=""
)

$extra_args_map = @{
    "timeout-seconds" = $timeout_seconds
    "comment" = $comment
    "output-s3-bucket-name" = $output_s3_bucket_name
    "output-s3-key-prefix" = $output_s3_key_prefix
    "service-role-arn" = $service_role_arn
    "notification-config" = $notification_config
    "cloud-watch-output-config" = $cloud_watch_output_config
    "alarm-configuration" = $alarm_configuration
    "endpoint-url" = $endpoint_url
    "region" = $region
}

# Validate configuration
function validate() {
    # Validate AWS CLI installation
    aws --version 2>&1 > $null
    if ($LASTEXITCODE -ne 0) {
        throw "AWS CLI was not found"
    }
}

# Build a string for extra arguments to pass to send-command api
function build_extra_args() {
    $extra_args = ""

    foreach ($arg in $extra_args_map.keys)
    {
        if (![string]::IsNullorWhitespace($extra_args_map[$arg])){
            $extra_args += "--$($arg) $($extra_args_map[$arg])"
        }
    }

    return $extra_args
}

# Helper function for send-command requests
function send_command() {
    Param(
        [parameter(Position=0,Mandatory=$true)]$instance_id,
        [parameter(Position=1,Mandatory=$true)]$target_os,
        [parameter(Position=2,Mandatory=$true)]$command,
        [parameter(Position=3,Mandatory=$false)]$extra_args
    )

    # Choose the SSM document to run based on the OS of the target instance
    if ($target_os -eq "windows"){
        $document_name = "AWS-RunPowerShellScript"
    }
    else {
        $document_name = "AWS-RunShellScript"
    }

    # Sanitize command
    $sanitized_command=($command -replace "'","\'" -replace "\$","``$" -replace "\\\\'","\\\'")

    # Execute SSM command on target instance
    $ssm_command=@"
    aws ssm send-command --instance-ids $($instance_id) --document-name $($document_name) --parameters commands="'$($sanitized_command)'" --output json $($extra_args)
"@
    Invoke-Expression $ssm_command
}

# Wait for SSM to be up and running on the target EC2 instance
function wait_ssm_bootstrap() {
    Param(
    [parameter(Position=0,Mandatory=$true)]$instance_id,
    [parameter(Position=1,Mandatory=$true)]$target_os,
    [parameter(Position=2,Mandatory=$true)]$ssm_timeout_seconds
    )

    $INVALID_STATE_ERR=@"
The instance is in an invalid state.
This can occur if:

- The instance ID is not valid
- The instance is not in the selected region
- The instance is not running
- The instance does not have the SSM agent running
- The instance does not have an instance profile with sufficient SSM permissions
"@

    # Wait for SSM to be up and running for a duration defined by the user
    $ssm_healthcheck_attempts=0
    $ssm_healthcheck_max_attempts=[math]::ceiling($ssm_timeout_seconds/10)

    while ($ssm_healthcheck_attempts -ne $ssm_healthcheck_max_attempts){
        Write-Output "Waiting for SSM to come online on instance $instance_id"
        $ssm_command_output=$(send_command -instance_id "$instance_id" -target_os "$target_os" -command "ls" 2>&1)
        if ($LASTEXITCODE -eq 0){
            Write-Output "SSM is online on instance $instance_id"
            return
        }
        else {
            Start-Sleep 10
            $ssm_healthcheck_attempts++
        }
    }

    Write-Output "SSM did not come online in the specified time"

    $stderr = ($ssm_command_output | Where-Object{ $_ -is [System.Management.Automation.ErrorRecord] } | out-string)

    # If the instance is not in a valid state, exit with a helpful message
    $instance_invalid_state = $stderr -match "Instances .* not in a valid state"
    if ($instance_invalid_state -eq $true) {
        Write-Output $INVALID_STATE_ERR
        exit 1
    } else {
        Write-Output "Something went wrong."
        Write-Output $stderr
        exit 1
    }
}

# Run SSM command on the target EC2 instance
function run_command(){
    Param(
    [parameter(Position=0,Mandatory=$true)]$instance_id,
    [parameter(Position=1,Mandatory=$true)]$target_os,
    [parameter(Position=2,Mandatory=$true)]$command,
    [parameter(Position=3,Mandatory=$true)]$wait_for_command_completion,
    [parameter(Position=4,Mandatory=$true)]$show_command_output,
    [parameter(Position=5,Mandatory=$true)]$continue_on_error
    )

    # Get extra args
    $extra_args=build_extra_args

    # Execute command and make sure a valid command id is returned
    $ssm_command_command_id=(send_command -instance_id "$instance_id" -target_os "$target_os" -command "$command" -extra_args "$extra_args" | ConvertFrom-Json).Command.CommandId

    if ($ssm_command_command_id -notmatch "^[a-z0-9-]+$") {
        Write-Output "Something went wrong."
        Write-Output "Output: $ssm_command_command_id"
        exit 1
    }

    # If wait_for_command_completion is false, exit immediately
    if ($wait_for_command_completion -eq "false") {
        Write-Output "SSM command has been executed"
        exit
    }

    # If a different region was specified, include it when retrieving ssm command details
    if (![string]::IsNullorWhitespace($extra_args_map["region"])){
        $region_arg = "--region $extra_args_map['region']"
    }

    # After the command has been executed, wait for its completion
    while ($true) {
        $ssm_get_command_invocation_command=@"
        aws ssm get-command-invocation --instance-id $($instance_id) --command-id $($ssm_command_command_id) $($region_arg)
"@
        $ssm_command_invocation=$(Invoke-Expression $ssm_get_command_invocation_command | ConvertFrom-Json)
        $ssm_command_invocation_stdout=$($ssm_command_invocation.StandardOutputContent )
        $ssm_command_invocation_stderr=$($ssm_command_invocation.StandardErrorContent )

        if (($ssm_command_invocation_stdout -eq "") -and ($ssm_command_invocation_stderr -eq "")){
            # If both STDOUT and STDERR are empty, then the command has not completed yet
        }
        else {
            if (($ssm_command_invocation_stdout -ne "") -and ($show_command_output -eq "true")){
                Write-Output "SSM command standard output: $ssm_command_invocation_stdout"
            }

            if ($ssm_command_invocation_stderr -ne ""){
                if ($show_command_output -eq "true"){
                    Write-Output "SSM command standard error: $ssm_command_invocation_stderr"
                }

                if ($continue_on_error -eq "false"){
                    throw "SSM command execution had errors"
                }
            }
            exit 0
        }
    }
}

# Validate configuration
validate

# Wait for SSM to be up and running on target instance
wait_ssm_bootstrap `
 -instance_id $instance_id `
 -target_os $target_os `
 -ssm_timeout_seconds $ssm_timeout_seconds

# Run SSM command on target instance
run_command `
  -instance_id $instance_id `
  -target_os $target_os `
  -command $command `
  -wait_for_command_completion $wait_for_command_completion `
  -show_command_output $show_command_output `
  -continue_on_error $continue_on_error